internal/plugin/admission/certificaterequest/approval/certificaterequest_approval.go

```diff
@@ -0,0 +1,285 @@
+		return resource, nil
+	}
+
+	// otherwise, query the apiserver
+	// TODO: we should enhance caching here to avoid performing discovery queries
+	//       many times if many CertificateRequest resources exist that reference
+	//       a resource that doesn't exist
+	groups, err := c.discovery.ServerGroups()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, apiGroup := range groups.Groups {
+		if apiGroup.Name != groupKind.Group {
+			continue
+		}
+
+		for _, version := range apiGroup.Versions {
+			apiResources, err := c.discovery.ServerResourcesForGroupVersion(version.GroupVersion)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, resource := range apiResources.APIResources {
+				if resource.Kind != groupKind.Kind {
+					continue
+				}
+
+				r := resource.DeepCopy()
+				// the Group field is not always populated in responses, so explicitly set it
+				r.Group = apiGroup.Name
+				c.cacheAPIResource(groupKind, *r)
+				return r, nil
+			}
+		}
+	}
+
+	return nil, errNoResourceExists
+}
+
+func (c *certificateRequestApproval) readAPIResourceFromCache(groupKind schema.GroupKind) *metav1.APIResource {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	if resource, ok := c.resourceCache[groupKind]; ok {
+		return &resource
+	}
+	return nil
+}
+
+func (c *certificateRequestApproval) cacheAPIResource(groupKind schema.GroupKind, resource metav1.APIResource) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.resourceCache[groupKind] = resource
+}
+
+var errNoResourceExists = fmt.Errorf("no resource registered")
+
+// signerNameForAPIResource returns the computed signerName for a given API resource
+// referenced by a CertificateRequest in a namespace.
+func signerNameForAPIResource(name, namespace string, apiResource metav1.APIResource) string {
+	if apiResource.Namespaced {
+		return fmt.Sprintf("%s.%s/%s.%s", apiResource.Name, apiResource.Group, namespace, name)
+	}
+	return fmt.Sprintf("%s.%s/%s", apiResource.Name, apiResource.Group, name)
+}
+
+// userInfoForRequest constructs a user.Info suitable for using with the authorizer interface
+// from an AdmissionRequest.
+func userInfoForRequest(req admissionv1.AdmissionRequest) user.Info {
+	extra := make(map[string][]string)
+	for k, v := range req.UserInfo.Extra {
+		extra[k] = v
+	}
+	return &user.DefaultInfo{
+		Name:   req.UserInfo.Username,
+		UID:    req.UserInfo.UID,
+		Groups: req.UserInfo.Groups,
+		Extra:  extra,
+	}
+}
+
+// isAuthorizedForSignerName checks whether an entity is authorized to 'approve' certificaterequests
+// for a given signerName.
+func isAuthorizedForSignerName(ctx context.Context, authz authorizer.Authorizer, info user.Info, signerName string) bool {
+	verb := "approve"
+	// First check if the user has explicit permission to 'approve' for the given signerName.
+	attr := buildAttributes(info, verb, signerName)
+	decision, _, err := authz.Authorize(ctx, attr)
+	switch {
+	case err != nil:
+		return false
```

Comment:
Should we not be bubbling up this error here? A network error would present itself as a permission denied.

Reply:
The authorizer itself performs retries to handle that particular case (and if I recall correctly, will actually retry indefinitely up to the configured request timeout). So I think this is okay (if there is a request timeout, the request won't be denied and the caller would retry the request (assuming it's a controller))

Reply:
Yep internal errors are retried indefinitely (using the configured backoff parameters in `internal/webhook`) so I think this is correct

Reply:
Cool, sgtm. Can we add a comment with this info to not snipe the next person? :smile: